api/hasura/cron/epochs.ts

```diff
@@ -725,6 +816,101 @@ async function notifyEpochStatus(
   return true;
 }
 
+function calculateNumReceived(
+  epoch: Pick<EpochsToNotify, 'endEpoch'>['endEpoch'][number]
+) {
+  return (epoch.circle?.users || []).reduce<
+    { email: string; tokenNumReceived: number; notesNumReceived: number }[]
+  >((acc, u) => {
+    const email = u.profile?.emails?.[0]?.email;
+    if (email) {
+      acc.push({
+        email,
+        tokenNumReceived: epoch.epoch_pending_token_gifts.reduce(
+          (count, gift) =>
+            count + (gift.recipient_id === u.id && gift.tokens > 0 ? 1 : 0),
+          0
+        ),
+        notesNumReceived: epoch.epoch_pending_token_gifts.reduce(
+          (count, gift) =>
+            count + (gift.recipient_id === u.id && gift.note ? 1 : 0),
+          0
+        ),
+      });
+    }
+    return acc;
+  }, []);
+}
+
+async function emailEpochStatus({
+  status,
+  circleId,
+  circleName,
+  membersData,
+}: {
+  status: 'started' | 'endingSoon' | 'ended';
+  circleId: number;
+  circleName: string;
+  membersData: Array<{
+    email: string;
+    tokenNumReceived?: number;
+    notesNumReceived?: number;
+  }>;
+}) {
+  try {
+    const responses = await Promise.allSettled(
+      status === 'ended'
+        ? membersData.map(memberData => {
+            return sendEpochEndedEmail({
+              email: memberData.email,
+              circle_id: circleId,
+              circle_name: circleName,
+              num_give_senders: memberData.tokenNumReceived ?? 0,
+              num_notes_received: memberData.notesNumReceived ?? 0,
+            });
+          })
+        : status === 'endingSoon'
+        ? membersData.map(memberData => {
+            return sendEpochEndingSoonEmail({
+              email: memberData.email,
+              circle_id: circleId,
+              circle_name: circleName,
+            });
+          })
+        : status === 'started'
+        ? membersData.map(memberData => {
+            return sendEpochStartedEmail({
+              email: memberData.email,
+              circle_id: circleId,
+              circle_name: circleName,
+            });
+          })
+        : []
+    );
+
+    const errors = responses?.filter(isRejected);
+    if (errors && errors.length > 0) {
+      const errorMessages = errors.map(err => {
+        if (err.reason instanceof Error) {
+          return err.reason.stack;
+        } else {
+          return String(err.reason);
+        }
+      });
+      console.error(
+        `Error sending email notifications: ${errorMessages.join('\n')}`
+      );
+      throw new Error(errorMessages.join('\n'));
+    }
+  } catch (e: unknown) {
+    // throwing here creates a promise rejection
+    if (e instanceof Error)
+      errorLog(`Error sending email notification: ${e.message}`);
```

Comment:
@sshmm maybe we should log any errors here with sentry? but lets do this in a follow up PR not in this PR